kernel/arch/riscv/process_riscv.c

```diff
@@ -91,57 +291,337 @@ void save_current_task_state(regs_t *r,  bool irq)
+    *
+    * For the moment, that is not necessary.
+    */
+   if (UNLIKELY(!are_interrupts_enabled())) {
+
+      /*
+       * Interrupts are disabled in this corner case: it's totally safe to read
+       * and write the static boolean.
+       */
+      if (!first_task_switch_passed) {
+
+         first_task_switch_passed = true;
+
+      } else {
+
+         /*
+          * Oops! We're not in the first task switch and interrupts are
+          * disabled: very likely there's a bug!
+          */
+         panic("Cannot switch away from task with interrupts disabled");
+      }
+   }
+
+   /*
+    * Make sure in NO WAY we'll switch to a user task keeping interrupts
+    * disabled. That would be a disaster. And if that happens due to a weird
+    * bug, let's try to learn as much as possible about why that happened.
+    */
+   if (UNLIKELY(!(next->state_regs->sstatus & SR_SIE) &&
+                !(next->state_regs->sstatus & SR_SPIE))) {
+
+      const char *curr_str =
+         curr->kthread_name
+            ? curr->kthread_name
+            : curr->pi->debug_cmdline;
+
+      const char *next_str =
+         next->kthread_name
+            ? next->kthread_name
+            : next->pi->debug_cmdline;
+
+      printk("[sched] task: %d (%p, %s) => %d (%p, %s)\n",
+             curr->tid, curr, curr_str,
+             next->tid, next, next_str);
+
+      if (next->running_in_kernel) {
+         dump_stacktrace(
+            regs_get_frame_ptr(next->state_regs),
+            next->pi->pdir
+         );
+      }
+
+      panic("[sched] Next task does not have interrupts enabled. "
+            "In kernel: %u, timer_ready: %u, is_sigsuspend: %u, "
+            "sa_pending: %p, sa_fault_pending: %p, "
+            "sa_mask: %p, sa_old_mask: %p",
+            next->running_in_kernel,
+            next->timer_ready,
+            next->in_sigsuspend,
+            next->sa_pending[0],
+            next->sa_fault_pending[0],
+            next->sa_mask[0],
+            next->sa_old_mask[0]);
+   }
 }
 
 NORETURN void
 switch_to_task(struct task *ti)
 {
-   NOT_IMPLEMENTED();
+   /* Save the value of ti->state_regs as it will be reset below */
+   regs_t *state = ti->state_regs;
+   struct task *curr = get_curr_task();
+
+   ASSERT(curr != NULL);
+
+   if (UNLIKELY(ti != curr)) {
+      ASSERT(curr->state != TASK_STATE_RUNNING);
+      ASSERT_TASK_STATE(ti->state, TASK_STATE_RUNNABLE);
+   }
+
+   ASSERT(!is_preemption_enabled());
+   switch_to_task_safety_checks(curr, ti);
+
+   /* Do as much as possible work before disabling the interrupts */
+   task_change_state_idempotent(ti, TASK_STATE_RUNNING);
+   ti->ticks.timeslice = 0;
+
+   if (!is_kernel_thread(curr) && curr->state != TASK_STATE_ZOMBIE)
+      save_curr_fpu_ctx_if_enabled();
+
+   if (!is_kernel_thread(ti)) {
+
+      if (get_curr_pdir() != ti->pi->pdir) {
+         set_curr_pdir(ti->pi->pdir);
+      }
+
+      if (!ti->running_in_kernel && !(state->sstatus & SR_SPP))
```

Comment:
I found here just check` ! ti->running_in_kernel` is not enough; On riscv, only `! (state->sstatus & SR_SPP)` indicates that you are about to switch to user space. I think the same problem may need to be fixed on the x86 architecture.

Reply:
@aenrbes Mmm.. that's interesting. Do you mind to elaborate a bit more? Is SR_SPP (Supervisor Previous Privilege) set by the CPU ? I have no idea how this works on RISCV.

On x86 we have IOPL should do the job because when switching to usermode IOPL must be 3. However, can you help me understand how `ti->running_in_kernel` can be false while we're going to switch to kernel mode? IRQs are non-preemptable and the syscall handling entry path disables the preemption before re-enabling the interrupts (see syscall_entry). Therefore, at least with a syscall I don't see how this could happen. (But that doesn't mean that I'm not missing anything!)

In case of anything else like an IRQ or a page fault, the `running_in_kernel` flag should not be set anyways..

Reply:
@vvaltchev 
When COW occurs from user space and `fault_entry()` is entered, `running_in_kernel` is not set at the time, and then when `enable_preemption()` in `fault_entry()` is executed, there may be a task switch. When it switch back here, `running_in_kernel` is zero but it returns to the kernel context.

Reply:
@aenrbes Ouch, that's looks like a problem. Can't we somehow call `set_current_task_in_kernel()` in `fault_entry()` to solve that?

```C
void fault_entry(regs_t *r)
{
   /*
    * Here preemption could be either enabled or disabled. Typically is enabled,
    * but it's totally possible for example a page fault to occur in the kernel
    * while preemption is disabled.
    */
   ASSERT(!are_interrupts_enabled());

   set_current_task_in_kernel();                // <<<<<<<<<< set running_in_kernel here
   push_nested_interrupt(regs_intnum(r));
   disable_preemption();
   enable_interrupts_forced();

   handle_fault(r);

   /*
    * Pop the nested interrupt and process signals while preemption is
    * still disabled.
    */
   pop_nested_interrupt();

   if (!get_curr_task()->running_in_kernel)
      process_signals(get_curr_task(), sig_in_fault, r);

   enable_preemption();
   set_current_task_in_user_mode();   //  <<<<<<<<<<<<<< disable it here
   disable_interrupts_forced();
}
```

**HOWEVER** , in case of a **nested fault** that would not work correctly. But if we make `running_in_kernel` to be an integer instead of a bool and do increase/decrease ? 

Not sure about the `irq_resched()` idea you mentioned. Even if we don't enable the preemption here, inside the fault handler the preemption can still be enabled. That's actually a pretty important use-case, and it's regular to get blocked and sleep in fault handlers in other OSs in order to do I/O and what not. 

Other OSs however don't allow page faults in the kernel due to CoW, but do `page_get()` and `page_put()` all the time. I was wondering if that can be avoided in Tilck, given the simplification that swap won't be supported. So if PF occurs, the page either was uncommitted or it's a CoW and requires a copy. In theory, this could work also with memory-mapped disk files, as long as the pages as always present. The I/O writes can be flushed to disk in an async fashion as long as the initial read happens on the first map. If I'm missing something else instead, we'd need to handle yet another case of PF where actual I/O is done. I was really trying hard to avoid all of that complexity in Tilck to not reimplement 100% of a classic *NIX kernel. I'm trying to keep it as simple as possible and yet.. it's never that simple once we have CoW, over-commitment, signals and what not :-( 

Anyway, do you think the idea of a ref-counted "in kernel" variable would work?

Reply:
I think a ref-counted "in kernel" variable is still problematic, because there are mechanisms in tilck which assume that if `running_in_kernel` is true, then it is represented in syscall, such as the one in `handle_potential_cow()`.

Reply:
The main problem is that the new "in kernel" variable needs to be able to distinguish between syscall and fault

Reply:
@aenrbes Mmm.. can you elaborate more on that? Tilck does not have cases where a page fault in the kernel is not triggered by a userspace process. So, if we get a page fault in handle_potential_cow(), it's OK to kill the app even if the page fault happened directly, instead of going through a syscall. Actually now it will work only in that case. If the PF happened in kernel while running a syscall, we'll panic the machine. 

Anyway, I'll think about that, but if you have a good idea in the meanwhile, please propose a fix.